pkg/api/validate/dynamic/dynamic.go

```diff
@@ -229,6 +237,55 @@ func (dv *dynamic) validateRouteTablePermissions(ctx context.Context, s Subnet)
 	return err
 }
 
+// validateNatGatewayPermissions will validate permissions on provided subnet
+func (dv *dynamic) validateNatGatewayPermissions(ctx context.Context, s Subnet) error {
+	dv.log.Printf("validateNatGatewayPermissions")
+
+	vnetID, _, err := subnet.Split(s.ID)
+	if err != nil {
+		return err
+	}
+
+	vnetr, err := azure.ParseResourceID(vnetID)
+	if err != nil {
+		return err
+	}
+
+	vnet, err := dv.virtualNetworks.Get(ctx, vnetr.ResourceGroup, vnetr.ResourceName, "")
+	if err != nil {
+		return err
+	}
+
+	ngID, err := getNatGatewayID(&vnet, s.ID)
+	if err != nil || ngID == "" { // error or no nat gateway
+		return err
+	}
+
+	ngr, err := azure.ParseResourceID(ngID)
+	if err != nil {
+		return err
+	}
+
+	errCode := api.CloudErrorCodeInvalidResourceProviderPermissions
+	if dv.authorizerType == AuthorizerClusterServicePrincipal {
+		errCode = api.CloudErrorCodeInvalidServicePrincipalPermissions
+	}
+
+	err = dv.validateActions(ctx, &ngr, []string{
+		"Microsoft.Network/natGateways/join/action",
+		"Microsoft.Network/natGateways/read",
+		"Microsoft.Network/natGateways/write",
+	})
+	if err == wait.ErrWaitTimeout {
```

Comment:
Can `wait.ErrWaitTimeout` happen for other issues other than lack of permissions? Usually a 403 is returned when the permissions don't line up and a timeout means we could not connect or the serving application did not respond in time

Reply:
This is a userspace wait timeout, which means that the loop inside validateActions (which reruns it because sometimes it takes a few goes if it's just been applied) didn't pass with success -- if the other end times out it'll be a HTTP error below.